010-server-side-rendering.md

```diff
@@ -0,0 +1,205 @@
+
+#### Netlify
+Netlify primarily focuses on static-site hosting. They also offer [Netlify Functions](https://www.netlify.com/products/functions/), a [user-friendly wrapper](https://docs.netlify.com/functions/overview/) around AWS Lambda. (These both support configuration [via Terraform](https://www.terraform.io/docs/providers/netlify/index.html), which is a huge plus in my book!)
+
+Unfortunately, their focus on static-site hosting means Netlify doesn't support on-demand server-side rendering out of the box (although there are somewhat hacky [ways to get it to work](https://medium.com/@finn.woelm/how-to-deploy-nextjs-on-netlify-with-server-side-rendering-9e8c06a06e77)). And like ZEIT Now, Netlify [requires that you use their CDN](https://www.netlify.com/blog/2017/03/28/why-you-dont-need-cloudflare-with-netlify/).
+
+In both of these cases, we'd be trading slightly higher costs (and reduced flexibility) for much simpler operations. This is a trade-off that we've happily made in the past (like our switch from EC2 to Heroku!), since we have a very small team & limited bandwidth for operations.
+
+However, being limited to a third party CDN is almost certainly a dealbreaker – it makes it impossible for us to gradually swap traffic to a new backend & would require re-writing our existing redirect logic (powered by Fastly's edge dictionaries) for a whole new stack.
+
+#### Heroku
+
+Because of those limitations, I'd recommend an initial rollout as a [traditional Node.js application](https://nextjs.org/docs/deployment#nodejs-server) on Heroku. This allows us to keep the same simple deployment infrastructure we're used to, while [running a parallel application with the new server-side renderer](https://elements.heroku.com/buildpacks/heroku/heroku-buildpack-multi-procfile). And while we'd be losing out on a lot of the scalability wins that are unlocked by moving to a serverless platform, there's nothing stopping us from making that additional move in the future.
+
+## Drawbacks
+
+While adding server-side rendering reduces context switching between two codebases (PHP and React), it introduces its own overhead since we can now render one codebase in two environments (Node.js for SSR & browser). Developers will need to be aware of these two differing "modes" a component may be rendered in, and re-write code appropriately (such as being careful about `window` usage).
+
+Server-side rendering is also still rather resource intensive, and we'll be heavily relying on caching to avoid overloading our backend. Luckily, we can always fall back to client-side rendering if necessary, and can use [`stale-while-revalidate`](https://www.fastly.com/blog/stale-while-revalidate-stale-if-error-available-today) to make slower requests to the backend without slowing down user-facing requests.
```

Comment:
`stale-while-revalidate` seems like a really neat idea!

> Luckily, we can always fall back to client-side rendering if necessary
Can you explain this bit a little more? I thought this was more of a switchover, how could we fall back?

Reply:
There should be nothing stopping any of these Next.js pages from being rendered client-side (and in fact, that's what's happening on subsequent navigation within the site). Server-side rendering is more of an enhancement where we save the client some work by shipping them a "finished product" the first time they visit a page!

So in the unlikely chance that server-side rendering was really taxing our servers, we could always "disable" it for a subset of pages & just ship the same sort of [application shell](https://github.com/DoSomething/phoenix-next/blob/master/resources/views/app.blade.php) (and render it all client-side). It'd be a more extreme version of the same way we'd opt-out of server-rendering some Apollo components with `ssr: false`.